enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/assets/k8s/PortForwards.java

```diff
@@ -0,0 +1,235 @@
+ *
+ * <p>Example Usage:</p>
+ * <pre>
+ * PortForwards manager = PortForwards.getDefault();
+ * manager.startPortForward(podItem);
+ * List<PortForwardItem> forwards = manager.getActivePortForwards("pod-name");
+ * manager.closePortForward("pod-name");
+ * </pre>
+ *
+ * @author Jan Horvath
+ */
+@NbBundle.Messages({
+    "PortNotFree=Port {0} is occupied by another process and cannot be used.",
+    "Forwarding=→ {0}"
+})
+public class PortForwards {
+
+    private static PortForwards instance = null;
+    private final Map<String, List<PortForwardItem>> activePortForwards;
+    private final Map<String, CountDownLatch> latches;
+    private final PropertyChangeSupport propertyChangeSupport;
+
+    private PortForwards() {
+        activePortForwards = new ConcurrentHashMap<>();
+        latches = new ConcurrentHashMap<>();
+        propertyChangeSupport = new PropertyChangeSupport(this);
+    }
+
+    public static synchronized PortForwards getDefault() {
+        if (instance == null) {
+            instance = new PortForwards();
+        }
+        return instance;
+    }
+    
+    /**
+     * Adds a {@code PropertyChangeListener} to listen for changes in port forwarding for a specific Pod.
+     *
+     * @param pod The {@code PodItem} for which the listener is interested.
+     * @param listener The listener to be added.
+     */
+    public void addPropertyChangeListener(PodItem pod, PropertyChangeListener listener) {
+        propertyChangeSupport.addPropertyChangeListener(pod.getName(), listener);
+    }
+
+    /**
+     * Removes a {@code PropertyChangeListener} for a specific Pod.
+     *
+     * @param pod The {@code PodItem} for which the listener should be removed.
+     * @param listener The listener to be removed.
+     */
+    public void removePropertyChangeListener(PodItem pod, PropertyChangeListener listener) {
+        propertyChangeSupport.removePropertyChangeListener(pod.getName(), listener);
+    }
+    
+    /**
+     * Notifies listeners of a property change for a specific Pod.
+     *
+     * @param pod The {@code PodItem} for which the change occurred.
+     * @param oldValue The old value of the property.
+     * @param newValue The new value of the property.
+     */
+    private void firePropertyChange(PodItem pod, Object oldValue, Object newValue) {
+        propertyChangeSupport.firePropertyChange(pod.getName(), oldValue, newValue);
+    }
+
+    /**
+     * Starts port forwarding for a specified Kubernetes Pod.
+     *
+     * <p>
+     * This method retrieves the Pod's ports and forwards them to the local
+     * machine. If any port is unavailable, the operation is halted for that
+     * port and a notification is displayed.</p>
+     *
+     * @param podItem The {@code PodItem} representing the Pod for which port
+     * forwarding should start.
+     */
+    public void startPortForward(PodItem podItem) {
+        if (activePortForwards.containsKey(podItem.getName())) {
```

Comment:
This does not prevent from starting portforward for the same Pod twice: the method may run, and will spawn a forward thread, again until `activePortForwards.put` after communicating with the k8s client.

If it was meant as sync/prevention of starting portforward twice, it should be fixed.